Disable srec tests.

--- ./ld/testsuite/ld-srec/srec.exp.aa	2014-08-04 10:28:16.313705173 -0700
+++ ./ld/testsuite/ld-srec/srec.exp	2014-08-04 10:29:30.839147853 -0700
@@ -391,6 +391,8 @@
 # generate the format if need be).
 setup_xfail "tile*-*-*"
 
+setup_xfail "powerpc64le*-*-*"
+
 run_srec_test $test1 "tmpdir/sr1.o tmpdir/sr2.o"
 
 # Now try linking a C++ program with global constructors and
@@ -421,5 +423,6 @@
 setup_xfail "score-*-*"
 setup_xfail "bfin-*-linux-uclibc"
 setup_xfail "tile*-*-*"
+setup_xfail "powerpc64le*-*-*"
 
 run_srec_test $test2 "tmpdir/sr3.o"
