diff -Nrup a/bfd/elf64-aarch64.c b/bfd/elf64-aarch64.c
--- a/bfd/elf64-aarch64.c	2015-05-28 11:22:51.726394256 -0600
+++ b/bfd/elf64-aarch64.c	2015-05-28 11:24:29.923395845 -0600
@@ -6006,17 +6006,6 @@ elf64_aarch64_reloc_type_class (const El
     }
 }
 
-/* Set the right machine number for an AArch64 ELF file.  */
-
-static bfd_boolean
-elf64_aarch64_section_flags (flagword *flags, const Elf_Internal_Shdr *hdr)
-{
-  if (hdr->sh_type == SHT_NOTE)
-    *flags |= SEC_LINK_ONCE | SEC_LINK_DUPLICATES_SAME_CONTENTS;
-
-  return TRUE;
-}
-
 /* Handle an AArch64 specific section when reading an object file.  This is
    called when bfd_section_from_shdr finds a section with an unknown
    type.  */
@@ -7944,9 +7933,6 @@ const struct elf_size_info elf64_aarch64
 #define elf_backend_reloc_type_class		\
   elf64_aarch64_reloc_type_class
 
-#define elf_backend_section_flags		\
-  elf64_aarch64_section_flags
-
 #define elf_backend_section_from_shdr		\
   elf64_aarch64_section_from_shdr
 
