--- binutils-2.23.52.0.1/bfd/elflink.c.~0~	2015-05-21 19:00:55.000000000 -0400
+++ binutils-2.23.52.0.1/bfd/elflink.c	2015-05-21 18:58:41.000000000 -0400
@@ -3803,7 +3803,7 @@
       /* We store a pointer to the hash table entry for each external
 	 symbol.  */
       amt = extsymcount * sizeof (struct elf_link_hash_entry *);
-      sym_hash = (struct elf_link_hash_entry **) bfd_alloc (abfd, amt);
+      sym_hash = (struct elf_link_hash_entry **) bfd_zalloc (abfd, amt);
       if (sym_hash == NULL)
 	goto error_free_sym;
       elf_sym_hashes (abfd) = sym_hash;
