commit c495064debc2cae93d0cb3710e4bf13bb9a0738a
Author: Andreas Schwab <schwab@linux-m68k.org>
Date:   Wed May 15 06:51:19 2013 +0000

    * elf64-aarch64.c (elf_backend_default_execstack): Define to 0.

diff --git a/bfd/elf64-aarch64.c b/bfd/elf64-aarch64.c
index e3cae95..aef472f 100644
--- a/bfd/elf64-aarch64.c
+++ b/bfd/elf64-aarch64.c
@@ -7181,6 +7181,7 @@ const struct elf_size_info elf64_aarch64_size_info =
 #define elf_backend_may_use_rela_p     1
 #define elf_backend_default_use_rela_p 1
 #define elf_backend_got_header_size (GOT_ENTRY_SIZE * 3)
+#define elf_backend_default_execstack  0
 
 #undef  elf_backend_obj_attrs_section
 #define elf_backend_obj_attrs_section		".ARM.attributes"
